group/assignment5/UnknownAlphabet.java

```diff
@@ -88,4 +91,74 @@
+     * @param dictionary list of strings
+     * @return minimal set of constrains that cannot be satisfied or empty Optional
+     */
+    public static Optional<Set<List<Character>>> checkInconsistency(List<String> dictionary) {
+        Map<Character, Set<Character>> contrains = new HashMap<>();
+        for (int i = 0; i < dictionary.size() - 1; i++) {
+            String w1 = dictionary.get(i);
+            String w2 = dictionary.get(i + 1);
+
+            for (int j = 0; j < Math.min(w1.length(), w2.length()); j++) {
+                if (w1.charAt(j) != w2.charAt(j)) {
+                    if (!contrains.containsKey(dictionary.get(i).charAt(j))) {
+                        // adds all the chars in the dictionary into the maps
+                        contrains.put(dictionary.get(i).charAt(j), new HashSet<>());
+                    }
+                    contrains.get(w1.charAt(j)).add(w2.charAt(j));
+                    break;
+                }
+            }
+        }
+        return checkForCircles(contrains);
+    }
+
+    /**
+     * Checks for any circles in the contrains
+     * @param contrains map ot the contrains
+     * @return minimal set of constrains that cannot be satisfied or empty Optional
+     */
+    private static Optional<Set<List<Character>>> checkForCircles(Map<Character, Set<Character>> contrains) {
+        for (Character before : contrains.keySet()) {
+            for (Character after : contrains.get(before)) {
+                Optional<List<Character>> path = findPath(contrains, after, before, new ArrayList<>(Arrays.asList(after)));
+                if (path.isPresent()) {
+                    Set<List<Character>> inconsistent = new HashSet<>();
+                    inconsistent.add(new ArrayList<>(Arrays.asList(before, after)));
+                    inconsistent.add(path.get());
+                    return Optional.of(inconsistent);
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Searches for a path in the contrains between from and to
+     * @param contrains map of the contrains
+     * @param from start node
+     * @param to end node
+     * @param path list, which represents the current path
+     * @return the complete path, if it was found; Optional.empty otherwise
+     */
+    private static Optional<List<Character>> findPath(Map<Character, Set<Character>> contrains, Character from, Character to, List<Character> path) {
+        if (from == to) {
+            return Optional.of(path);
+        }
+        if (contrains.containsKey(from)) {
+            for (Character after : contrains.get(from)) {
+                path.add(after);
+                Optional<List<Character>> foundPath = findPath(contrains, after, to, path);
+                if (foundPath.isPresent()) {
+                    return foundPath;
+                }
+                path.remove(after);
+            }
+        }
+        return Optional.empty();
+    }
```

Comment:
This comment is more of a general one:
- I implemented the challenge 3 on top of your current code (since it was requiring challenge 2 to solve it). When we do some changes in here and try to merge both parts we might have conflict and we might need to solve it but I think it would be fairly easy since I didn't touch at all the part you implemented for challenge 2 :) I just wanted to comment here to make you aware about such a possibility :D